File: internal/wallet/profile_store.go
package wallet

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"
)

const (
	profileLoadRetries = 3
	profileRetryDelay  = 25 * time.Millisecond
)

type ProfileStore struct {
	profileDir string
	cache      *profileStoreCache
}

func NewProfileStore(profileDir string) *ProfileStore {
	cleanedDir := filepath.Clean(profileDir)
	return &ProfileStore{
		profileDir: cleanedDir,
		cache:      sharedProfileStoreCache(cleanedDir),
	}
}

func (store *ProfileStore) Load(profileName string) (*PlayerProfileState, error) {
	if cached, ok := store.cache.load(profileName); ok {
		return cached, nil
	}

	path := store.pathFor(profileName)
	for attempt := 0; attempt < profileLoadRetries; attempt++ {
		data, err := os.ReadFile(path)
		if err != nil {
			if errors.Is(err, os.ErrNotExist) {
				return nil, nil
			}
			return nil, err
		}
		if len(data) == 0 {
			if attempt < profileLoadRetries-1 {
				time.Sleep(profileRetryDelay)
				continue
			}
			return nil, nil
		}

		var state PlayerProfileState
		if err := json.Unmarshal(data, &state); err != nil {
			if attempt < profileLoadRetries-1 {
				time.Sleep(profileRetryDelay)
				continue
			}
			return nil, err
		}
		store.cache.save(state)
		return &state, nil
	}
	return nil, nil
}

func (store *ProfileStore) Save(state PlayerProfileState) error {
	if err := os.MkdirAll(store.profileDir, 0o755); err != nil {
		return err
	}

	path := store.pathFor(state.ProfileName)
	data, err := json.MarshalIndent(state, "", "  ")
	if err != nil {
		return err
	}
	data = append(data, '\n')

	if err := writeFileAtomic(path, data, 0o644); err != nil {
		return err
	}
	store.cache.save(state)
	return nil
}

func (store *ProfileStore) ListProfileNames() ([]string, error) {
	entries, err := os.ReadDir(store.profileDir)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}

	names := make([]string, 0, len(entries))
	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".json") {
			continue
		}
		names = append(names, strings.TrimSuffix(entry.Name(), ".json"))
	}
	sort.Strings(names)
	return names, nil
}

func (store *ProfileStore) LoadSummary(profileName string) (*LocalProfileSummary, error) {
	state, err := store.Load(profileName)
	if err != nil || state == nil {
		return nil, err
	}
	summary := ToLocalProfileSummary(*state)
	return &summary, nil
}

func (store *ProfileStore) ListProfiles() ([]LocalProfileSummary, error) {
	names, err := store.ListProfileNames()
	if err != nil {
		return nil, err
	}

	summaries := make([]LocalProfileSummary, 0, len(names))
	for _, name := range names {
		state, err := store.Load(name)
		if err != nil {
			return nil, err
		}
		if state == nil {
			continue
		}
		summaries = append(summaries, ToLocalProfileSummary(*state))
	}
	return summaries, nil
}

func (store *ProfileStore) pathFor(profileName string) string {
	return filepath.Join(store.profileDir, profileName+".json")
}

type profileStoreCache struct {
	mu       sync.RWMutex
	profiles map[string]PlayerProfileState
}

var sharedProfileStoreCaches sync.Map

func sharedProfileStoreCache(profileDir string) *profileStoreCache {
	if cache, ok := sharedProfileStoreCaches.Load(profileDir); ok {
		return cache.(*profileStoreCache)
	}
	cache := &profileStoreCache{profiles: map[string]PlayerProfileState{}}
	actual, _ := sharedProfileStoreCaches.LoadOrStore(profileDir, cache)
	return actual.(*profileStoreCache)
}

func (cache *profileStoreCache) load(profileName string) (*PlayerProfileState, bool) {
	cache.mu.RLock()
	state, ok := cache.profiles[profileName]
	cache.mu.RUnlock()
	if !ok {
		return nil, false
	}
	return clonePlayerProfileState(&state), true
}

func (cache *profileStoreCache) save(state PlayerProfileState) {
	cloned := clonePlayerProfileState(&state)
	cache.mu.Lock()
	cache.profiles[state.ProfileName] = *cloned
	cache.mu.Unlock()
}

func writeFileAtomic(path string, data []byte, mode os.FileMode) error {
	tempFile, err := os.CreateTemp(filepath.Dir(path), filepath.Base(path)+".*.tmp")
	if err != nil {
		return err
	}
	tempPath := tempFile.Name()
	defer os.Remove(tempPath)

	if _, err := tempFile.Write(data); err != nil {
		tempFile.Close()
		return err
	}
	if err := tempFile.Chmod(mode); err != nil {
		tempFile.Close()
		return err
	}
	if err := tempFile.Close(); err != nil {
		return err
	}
	return os.Rename(tempPath, path)
}

func ToLocalProfileSummary(profile PlayerProfileState) LocalProfileSummary {
	summary := LocalProfileSummary{
		HasPeerIdentity:      profile.PeerPrivateKeyHex != "",
		HasProtocolIdentity:  profile.ProtocolPrivateKeyHex != "",
		HasTransportIdentity: profile.TransportPrivateKeyHex != "",
		HasWalletIdentity:    profile.WalletPrivateKeyHex != "",
		KnownPeerCount:       len(profile.KnownPeers),
		MeshTableCount:       len(profile.MeshTables),
		Nickname:             profile.Nickname,
		ProfileName:          profile.ProfileName,
	}
	if profile.CurrentMeshTableID != "" {
		summary.CurrentMeshTableID = profile.CurrentMeshTableID
	}
	if profile.CurrentTable != nil {
		summary.CurrentTableID = profile.CurrentTable.TableID
	}
	return summary
}

File: packages/daemon-runtime/src/profileStore.ts
import { mkdir, readdir, readFile, rename, writeFile } from "node:fs/promises";
import { join } from "node:path";

import type { MeshTableConfig } from "@parker/protocol";
import type { WalletSummary } from "@parker/settlement";

export interface TableSessionState {
  inviteCode: string;
  seatIndex: 0 | 1;
  tableId: string;
}

export interface KnownPeerState {
  alias?: string;
  lastSeenAt?: string;
  peerId: string;
  peerUrl: string;
  protocolPubkeyHex?: string;
  relayPeerId?: string;
  roles?: Array<"player" | "host" | "witness" | "indexer">;
}

export interface MeshTableReferenceState {
  config?: MeshTableConfig;
  currentEpoch: number;
  hostPeerId: string;
  hostPeerUrl: string;
  role: "host" | "player" | "witness";
  tableId: string;
  visibility: "private" | "public";
}

export interface PlayerProfileState {
  currentTable?: TableSessionState;
  currentMeshTableId?: string;
  handSeeds: Record<string, string>;
  knownPeers?: KnownPeerState[];
  meshTables?: Record<string, MeshTableReferenceState>;
  mockWallet?: WalletSummary;
  nickname: string;
  peerPrivateKeyHex?: string;
  privateKeyHex: string;
  protocolPrivateKeyHex?: string;
  profileName: string;
  walletPrivateKeyHex?: string;
}

export interface LocalProfileSummary {
  currentMeshTableId?: string;
  currentTableId?: string;
  hasPeerIdentity: boolean;
  hasProtocolIdentity: boolean;
  hasWalletIdentity: boolean;
  knownPeerCount: number;
  meshTableCount: number;
  nickname: string;
  profileName: string;
}

const PROFILE_LOAD_RETRY_MS = 25;
const PROFILE_LOAD_RETRIES = 3;
const profileWriteLocks = new Map<string, Promise<void>>();

export class ProfileStore {
  constructor(private readonly profileDir: string) {}

  async listProfileNames(): Promise<string[]> {
    try {
      const entries = await readdir(this.profileDir, { withFileTypes: true });
      return entries
        .filter((entry) => entry.isFile() && entry.name.endsWith(".json"))
        .map((entry) => entry.name.slice(0, -".json".length))
        .sort((left, right) => left.localeCompare(right));
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === "ENOENT") {
        return [];
      }
      throw error;
    }
  }

  async listProfiles(): Promise<LocalProfileSummary[]> {
    const profiles = await Promise.all(
      (await this.listProfileNames()).map(async (profileName) => {
        const profile = await this.load(profileName);
        return profile ? toLocalProfileSummary(profile) : null;
      }),
    );
    return profiles.filter((profile): profile is LocalProfileSummary => profile !== null);
  }

  async loadSummary(profileName: string): Promise<LocalProfileSummary | null> {
    const profile = await this.load(profileName);
    return profile ? toLocalProfileSummary(profile) : null;
  }

  async load(profileName: string): Promise<PlayerProfileState | null> {
    for (let attempt = 0; attempt < PROFILE_LOAD_RETRIES; attempt += 1) {
      try {
        const raw = await readFile(this.pathFor(profileName), "utf8");
        if (!raw.trim()) {
          if (attempt < PROFILE_LOAD_RETRIES - 1) {
            await this.sleep(PROFILE_LOAD_RETRY_MS);
            continue;
          }
          return null;
        }
        return JSON.parse(raw) as PlayerProfileState;
      } catch (error) {
        if ((error as NodeJS.ErrnoException).code === "ENOENT") {
          return null;
        }
        if (error instanceof SyntaxError && attempt < PROFILE_LOAD_RETRIES - 1) {
          await this.sleep(PROFILE_LOAD_RETRY_MS);
          continue;
        }
        throw error;
      }
    }
    return null;
  }

  async save(state: PlayerProfileState): Promise<void> {
    await mkdir(this.profileDir, { recursive: true });
    const path = this.pathFor(state.profileName);
    const previousWrite = profileWriteLocks.get(path) ?? Promise.resolve();
    const nextWrite = previousWrite.catch(() => undefined).then(async () => {
      const tempPath = `${path}.${process.pid}.${crypto.randomUUID()}.tmp`;
      try {
        await writeFile(tempPath, `${JSON.stringify(state, null, 2)}\n`, "utf8");
        await rename(tempPath, path);
      } catch (error) {
        if ((error as NodeJS.ErrnoException).code === "ENOENT") {
          return;
        }
        throw error;
      }
    });
    profileWriteLocks.set(path, nextWrite);
    try {
      await nextWrite;
    } finally {
      if (profileWriteLocks.get(path) === nextWrite) {
        profileWriteLocks.delete(path);
      }
    }
  }

  private pathFor(profileName: string) {
    return join(this.profileDir, `${profileName}.json`);
  }

  private async sleep(timeoutMs: number) {
    await new Promise<void>((resolve) => {
      setTimeout(resolve, timeoutMs);
    });
  }
}

function toLocalProfileSummary(profile: PlayerProfileState): LocalProfileSummary {
  return {
    ...(profile.currentMeshTableId ? { currentMeshTableId: profile.currentMeshTableId } : {}),
    ...(profile.currentTable ? { currentTableId: profile.currentTable.tableId } : {}),
    hasPeerIdentity: Boolean(profile.peerPrivateKeyHex),
    hasProtocolIdentity: Boolean(profile.protocolPrivateKeyHex),
    hasWalletIdentity: Boolean(profile.walletPrivateKeyHex),
    knownPeerCount: profile.knownPeers?.length ?? 0,
    meshTableCount: Object.keys(profile.meshTables ?? {}).length,
    nickname: profile.nickname,
    profileName: profile.profileName,
  };
}
